### ISSUE TITLE
PBFT goes to viewchange after a few transactions

### ISSUE THREAD

**Issue by esgott** [hyperledger-archives/fabric#1741]
## Description

I have a PBFT network a 4 nodes, and trying to measure the performance I could get out of it. Unfortunately, the network always go to view-change, and never (I waited about a minute) succeeds to go to a new view.. First I thought, that it may happen because of the big stress of the transactions sent in, so I put a 1 sec timer between transactions, but I get the same problem. Then I thought, maybe because of short timers, so I changed the obcpbft config to have 10 sec instead of 2 sec, but that did not help either. I attach the logs of the peers.
[pbft_problem.zip](https://github.com/hyperledger/fabric/files/302562/pbft_problem.zip)
## Describe How to Reproduce

4 t2.large nodes on EC2, running hyperledger-peer docker image.
I use [PR 1722](https://github.com/hyperledger/fabric/pull/1722) and [PR 1721](https://github.com/hyperledger/fabric/pull/1721) on top of master, my assumption is, that it should not affect PBFT consensus.
I run [this](https://github.com/esgott/fabric-api/blob/perf-test/api/src/test/java/org/hyperledger/connector/PerfTest.java) for performance testing.
In configuration, I change the following things from default:
- `peer.id:` vp0, vp1, vp2 and vp3
- `peer.address:` EC2_instance_private_address:30303
- `peer.validator.consensus.plugin:` pbft
- `peer.discovery.rootnode:` vp0_private_address:30303

**Comment by corecode** [comment]
The network stops working because the peers started out at block=110 with different state:

```
vp0: cset:<sequence_number:110 id:"CG8SQEODtq4R9rx+s/QNau8NfH7Hfej8OJE8GW41J5DzFOXppPjoWOSGSccSsjp0RIrDDj8Wm0eJR+dJLikUVQZ1paYaQBQ8cEC4oUH0imdsXHico3bh/hmYACviyPgnQQISf5m9lRJGh1DL2hsFOKB5Zdk6psD5grXnByk5HPUaAh1pm+Y=" >
vp1: cset:<sequence_number:110 id:"CG8SQLxe7qHuU5QJXUZcetQ9LDxgFd/bkWmsqtBLTMJJuO1l3NfZczmzqcR1pRPLzmn7mN0SeDahDVX/Ke9G6C6/d6QaQKJFgGfAZTwe+o5XNvd+B067Myl2M4Tbi6eQYDuZ7wZPS6TskK3vDDQlnHIq3QtEnekSULZNsy4pGgo/ARCGfnc=" >
vp3: cset:<sequence_number:110 id:"CG8SQIKDEEthOW4a4pRHyI/EhOmYsXB4Wivqa8bm5YCuXFPtaVyyp/CGkcWYBqCbqUgyLYOI2X5ZHBLQJJKAsmEvWuMaQFgwvcT2hMgVqMjTkH0XCw04NJQK+EbdoAlWNoFRZKxlx4Es5b7PNnAPlACXRsVxjENjW53CyMBxI9qKqMTpVxo=" >
```

The network stopping operation in this situation is an expected and reasonable outcome.

**Comment by vukolic** [comment]
@esgott Arriving to state divergence in block 110 - did you issue some queries concurrently with invokes?

**Comment by esgott** [comment]
@corecode I think the question is then why are the peers in a different state? I am just feeding one of them with transactions.

@vukolic no concurrent queries, at least not that I know of. But even if there would be, shouldn't it be handled by the fabric peers?

**Comment by corecode** [comment]
@esgott I don't know.  You started the network.  Did you maybe use noops at some point on the same ledger?

**Comment by esgott** [comment]
@corecode no

**Comment by tamasblummer** [comment]
@corecode why is stopping the network a reasonable outcome for discovering a misbehaving peer?

**Comment by corecode** [comment]
@tamasblummer more than F replicas diverged.  There are no more guarantees the network can make.  We cannot continue.

**Comment by tamasblummer** [comment]
@corecode F is 1 in our case, so we have more than 1 misbehaving peers although all run the same code and only one receives transactions. That is not expected.

**Comment by corecode** [comment]
@tamasblummer The logs you provided show that you _started_ the network already in divergent state.  I don't know how you got there, but for the logs you provided, the behavior is absolutely correct.

I'm not saying there are no bugs or unhandled corner cases in the code.  However, I will need the complete logs that show how the divergence happened in the first place.

**Comment by tamasblummer** [comment]
@corecode what needs to be deleted to ensure this does not happen?

**Comment by corecode** [comment]
Please erase the whole ledger database in `/var/hyperledger`

**Comment by vukolic** [comment]
@esgott I am asking because of potential link to #1751 

It would be helpful if you could provide some history of your system up to seqNo 110

**Comment by corecode** [comment]
Thank you.  The issue seems similar.  Could you please try with batch consensus, because classic is deprecated and will be removed soon.  If the problem persists, please supply logs in debug level (rest can stay at warn).

**Comment by esgott** [comment]
@corecode which group would you like to see on debug level?

**Comment by corecode** [comment]
`consensus/obcpbft`, and maybe `consensus/obcpbft/custodian`

**Comment by esgott** [comment]
@corecode 
[pbft_problem_3.zip](https://github.com/hyperledger/fabric/files/305084/pbft_problem_3.zip)

**Comment by corecode** [comment]
Unfortunately this doesn't look like debug level for obcpbft

**Comment by esgott** [comment]
I started the peer with this (inside docker):
`rm -rf /var/hyperledger && OPENCHAIN_LOGGING_PEER=info,consensus/obcpbft=debug peer node start &> output.txt`

What would be the correct way?

**Comment by corecode** [comment]
my mistake - i guess by now it should be something with hyperledger or core...

**Comment by corecode** [comment]
`CORE_LOGGING_PEER`?

**Comment by esgott** [comment]
@corecode I've tried with CORE_LOGGING_PEER=info,consensus/obcpbft=debug, still no debug messages for obcpbft

**Comment by esgott** [comment]
@corecode I managed to turn it on, with CORE_LOGGING_LEVEL.
[pbft_problem_4.zip](https://drive.google.com/file/d/0B3kmt1BIHlg4VGxVMGY5VmlpLUk/view?usp=sharing)

**Comment by corecode** [comment]
thanks!

**Comment by corecode** [comment]
What system chaincode are you using?

**Comment by esgott** [comment]
noop, as in [PR 1722](https://github.com/hyperledger/fabric/pull/1722)

**Comment by corecode** [comment]
Thank you.  Somehow the checkpoints show different block hashes.  Can you run the whole system in debug, so that we can trace the block creation?

**Comment by esgott** [comment]
[pbft_problem_5.zip](https://github.com/hyperledger/fabric/files/306920/pbft_problem_5.zip)

**Comment by corecode** [comment]
seems most log levels are still on INFO

**Comment by esgott** [comment]
Hope it's better
[pbft_problem_6.zip](https://github.com/hyperledger/fabric/files/306949/pbft_problem_6.zip)

**Comment by corecode** [comment]
I have no idea why ledger or buckettree are not providing any output.  Maybe @srderson or @manish-sethi can help?

**Comment by jyellick** [comment]
Just to chime in in support of @corecode's analysis, as far as the PBFT logs indicate, all of the same requests are being executed in the exact same order on every peer, but for some reason the state seems to diverge across them.  The lack of ledger logs is also puzzling, so some help from the ledger folks above would be appreciated.

**Comment by corecode** [comment]
I figured out that setting the logging level is done with `CORE_LOGGING_LEVEL=debug` or `CORE_LOGGING_NODE=debug` - seems the naming got changed at some point.

**Comment by manish-sethi** [comment]
I was also looking at the code for enabling logging. logging for selective modules can be enable for debug by a global flag --logging-level. 
e.g., `--logging-level=ledger,state,buckettree,chaincode,consensus/obcpbft=debug` would enable debug messages for these modules.

**Comment by corecode** [comment]
@muralisrini I found the problem.  Deployed system chaincode does not produce the same genesis block across all nodes.

**Comment by manish-sethi** [comment]
@corecode so it was block hash that was different not the stateHash - is this correct? Also, why does the block hash differ at block 110 if the genesis blocks were different?

**Comment by esgott** [comment]
@muralisrini noop sys chaincode, as in [PR 1722](https://github.com/hyperledger/fabric/pull/1722)

**Comment by corecode** [comment]
This is the problem:
https://github.com/hyperledger/fabric/blob/master/core/ledger/genesis/genesis.go#L156
https://github.com/hyperledger/fabric/blob/master/core/ledger/genesis/genesis.go#L221
https://github.com/hyperledger/fabric/blob/master/protos/transaction.go#L70

**Comment by corecode** [comment]
Also, is the iteration of genesis chaincodes a stable sequence, or is that map (random) sequence?

**Comment by esgott** [comment]
@corecode Do I understand correctly, that the problem is the following:
During startup, all the nodes will deploy the noop system chaincode, but all of them will make a different transaction for it, so the network starts up with 4 forks of the blockchain?

**Comment by corecode** [comment]
yes.

**Comment by binhn** [comment]
Sorry folks.  @muralisrini and I talked about this potential issue weeks ago, and fixed it in another branch with the "uber" system chaincode, but uber hasn't made it in. 

In hindsight, we should have just fixed this issue independently so that not so many people wasted so much time.

The fix was to remove the system chaincodes from the genesis block and treat them as part of peer on deployment. @muralisrini will add more details with the pull request.

@esgott @tamasblummer Thanks for the test case !

**Comment by esgott** [comment]
@binhn Thanks for the info. Can you update this thread when it's ready, and close this afterwards?

**Comment by muralisrini** [comment]
@esgott will update here with an approach for further comments/review.

**Comment by muralisrini** [comment]
@here description/fix in #1826
